src/plotter_disk.hpp

```diff
@@ -73,8 +76,639 @@ struct Phase3Results {
+
+    // Start at left table pos = 0 and iterate through the whole table. Note that the left table
+    // will already be sorted by y
+    uint64_t totalstripes = (prevtableentries + STRIPESIZE - 1) / STRIPESIZE;
+    uint64_t threadstripes = (totalstripes + NUMTHREADS - 1) / NUMTHREADS;
+
+    for (uint64_t stripe = 0; stripe < threadstripes; stripe++) {
+        uint64_t pos = (stripe * NUMTHREADS + ptd->index) * STRIPESIZE;
+        uint64_t endpos = pos + STRIPESIZE + 1;  // one y value overlap
+        uint64_t left_reader = pos * entry_size_bytes;
+        uint64_t left_reader_count = 0;
+        uint64_t left_writer_count = 0;
+        uint64_t stripe_left_writer_count = 0;
+        uint64_t stripe_start_correction = 0xffffffffffffffff;
+        uint64_t right_writer_count = 0;
+        uint64_t matches = 0;  // Total matches
+
+        // This is a sliding window of entries, since things in bucket i can match with things in
+        // bucket
+        // i + 1. At the end of each bucket, we find matches between the two previous buckets.
+        std::vector<PlotEntry> bucket_L;
+        std::vector<PlotEntry> bucket_R;
+
+        uint64_t bucket = 0;
+        bool end_of_table = false;  // We finished all entries in the left table
+
+        uint64_t ignorebucket = 0xffffffffffffffff;
+        bool bMatch = false;
+        bool bFirstStripeOvertimePair = false;
+        bool bSecondStripOvertimePair = false;
+        bool bThirdStripeOvertimePair = false;
+
+        bool bStripePregamePair = false;
+        bool bStripeStartPair = false;
+
+        uint64_t L_position_base = 0;
+        uint64_t R_position_base = 0;
+        uint64_t newlpos = 0;
+        uint64_t newrpos = 0;
+        Bits new_left_entry(0, pos_size + kOffsetSize);
+        std::vector<std::tuple<PlotEntry, PlotEntry, std::pair<Bits, Bits>>>
+            current_entries_to_write;
+        std::vector<std::tuple<PlotEntry, PlotEntry, std::pair<Bits, Bits>>>
+            future_entries_to_write;
+        std::vector<std::pair<uint16_t, uint16_t>> match_indexes;
+        std::vector<PlotEntry*> not_dropped;  // Pointers are stored to avoid copying entries
+
+        if (pos == 0) {
+            bMatch = true;
+            bStripePregamePair = true;
+            bStripeStartPair = true;
+            stripe_left_writer_count = 0;
+            stripe_start_correction = 0;
+        }
+
+#ifdef _WIN32
+        WaitForSingleObject(ptd->theirs, INFINITE);
+#else
+        sem_wait(ptd->theirs);
+#endif
+        if (pos < prevtableentries + 1) {
+            uint64_t readamt = std::min(
+                ((uint64_t)(STRIPESIZE) + 2500) * entry_size_bytes,
+                ((prevtableentries + 1) * entry_size_bytes) - left_reader);
+
+            (*ptmp_1_disks)[table_index].Read(left_reader, left_reader_buf, readamt);
+            left_reader_count = 0;
+        }
+#ifdef _WIN32
+        ReleaseSemaphore(ptd->mine, 1, NULL);
```

Comment:
it's not clear to me why the threads wait for each other (in order), and then perform some serial work, and then kick the next thread to do some serial work. You have this cascade of semaphores to implement this.

Wouldn't it be simpler if all the parallel work happens first, then all threads join and then all the serial work is performed? Is the serial work so expensive there's a material gain from allowing threads ahead of it to keep working on the parallel work while one thread performs the serial work?

Reply:
Hi arvid

Yes we are cascading semaphores here to order things. The workload is divided into small stripes as we don't have enough memory to do the whole thing. The stripesize is configurable though, so if you had enough memory you could divide it evenly as you have said.  We have to serialize things because the stripe can only calculate the relative position within the stripe but the table demands the absolute position. So we have to correct things serially as the table is written out. There are other reasons for serializing things as well (limiting random disk seeks eg). Currently the code does not scale up very well past a few CPUs though so certainly lots of room for improvement here. The entry dropping code demands a lot of synchronization but it is worth it for the typical user. ramwjb runs fastest (48 cpu load) but all ram is pretty specialized. Let me know if you have any ideas!!!!

Reply:
BTW there is a video in keybase about the entry dropping and striping: (striping-bill)

Reply:
I see, what I had missed was that the thread continues the "parallel work" again. I first had read it as it was doing the serial work at the end, right before the thread would finish, and return from the thread function.